aws_eks_cluster: Add wait functionality

deiwin opened the pull request:
##### SUMMARY
Add option to wait until cluster is fully created and active. This enables further cluster configuration once it's created and active.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
aws_eks_cluster

##### ANSIBLE VERSION
2.7

##### ADDITIONAL INFORMATION
Some additional info in commit messages.

deiwin committed:
aws_eks_cluster: Improve output documentation

This data is already returned by the module, it just wasn't documented. These
fields are required for accessing the created Kubernetes API with e.g. the
k8s_raw module.

deiwin committed:
aws_eks_cluster: Add wait functionality

This enables further cluster configuration once it's created and active.

20 minutes was chosen as an arbitrary default, so that if it takes longer than
the documented "usually less than 10 minutes" it's still likely to succeed.

ryansb reviewed (commented):
I haven't tested the wait functionality on the module, but the waiter itself looks good.

willthames commented on lib/ansible/modules/cloud/amazon/aws_eks_cluster.py:
```diff
@@ -36,6 +36,18 @@
+    description: >-
+      The duration in seconds to wait for the cluster to become active. Defaults
+      to 1200 seconds (20 minutes).
+    type: int
```
This line is failing CI - just remove it (it makes no sense that `type: int` isn't valid here, but it isn't)

deiwin replied:
Right. Removed.

deiwin commented:
Maybe I'm looking at it wrong, but it seems to me that Shippable is running some `ec2_vpc_vgw` tests, for some reason, and I don't see the EKS test anywhere.

willthames commented:
The aws_eks tests are marked as unsupported (test/integration/aws_eks/aliases), and so don't run automatically.

I'll run these myself manually.

deiwin commented:
Okay, thanks.

willthames committed:
Correct security group name in aws_eks tests

willthames commented:
I've added a tiny change to this commit as the test suite wasn't cleaning up after itself properly. This is a fix for an issue I introduced that I thought had been resolved before the module was merged last time.

I'll squash and merge once shippable tests pass

willthames commented:
Tests still aren't passing I'm afraid - I've found another SG issue, but we might need a waiter for delete to make the test suite pass.

deiwin commented:
Do you mean a waiter for waiting for an EKS cluster to get deleted with `state: absent`?

willthames commented on test/integration/targets/aws_eks/playbooks/roles/aws_eks/tasks/main.yml:
```diff
@@ -106,6 +106,27 @@
+          - eks_create.certificate_authority.data is defined
+          - eks_create.certificate_authority.data is not ""
+          - eks_create.endpoint is defined
+          - eks_create.endpoint is not ""
```
Should be `!=` rather than `is not` (currently tests fail here)

deiwin replied:
Updated this and the other similar issue.

willthames commented on test/integration/targets/aws_eks/playbooks/roles/aws_eks/tasks/main.yml:
```diff
@@ -106,6 +106,27 @@
+          - eks_create.name == eks_cluster_name
+          - eks_create.status == "ACTIVE"
+          - eks_create.certificate_authority.data is defined
+          - eks_create.certificate_authority.data is not ""
```
Should be != rather than is not (currently tests fail here)

willthames commented:
@deiwin - a waiter that waits for ResourceNotFound, I would think, when `state: absent`, as you suggest. It's not a blocker for this PR though as it's unrelated to your changes, they just expose an existing bug, I'll try with a `pause: minutes=5` for now.

willthames committed:
Improve teardown of aws_eks tests

Fix minor teardown issues. The `pause` step is a placeholder until
a waiter for `state: absent`

willthames commented:
Tests are now passing at my end with a couple more tweaks, I'll merge this once shippable passes.

willthames commented:
Merged, thanks @deiwin

deiwin commented:
Thanks!
